File: app-mobile/lib/domain/appointment/appointment.dart
// Entité de domaine « rendez-vous » côté client (domaine réservation, #22).
//
// Domaine **pur** : aucune dépendance à Flutter ni à un client HTTP (ADR-0008).
// Reflète l'`AppointmentResponse` renvoyée par `POST /salons/{id}/appointments`
// (#21) au statut `PENDING`. Ne porte que **les données du RDV du client** —
// jamais l'identité d'autres clients ni de donnée de gestion (§11.3).

import 'appointment_status.dart';

/// Prestation réservée : identifiant + prix **figé** au moment de la réservation.
class BookedService {
  const BookedService({required this.serviceId, this.priceAtBooking});

  /// Identifiant opaque de la prestation (UUID côté backend).
  final String serviceId;

  /// Prix figé (chaîne décimale, p. ex. « 5000.00 ») ou `null` si non fourni.
  /// Conservé en chaîne pour ne pas introduire d'imprécision flottante.
  final String? priceAtBooking;
}

/// Rendez-vous créé, tel que renvoyé par le backend à la réservation.
class Appointment {
  const Appointment({
    required this.id,
    required this.salonId,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.status,
    this.hairdresserId,
    this.clientNote,
    this.services = const <BookedService>[],
  });

  /// Identifiant opaque du RDV (UUID côté backend).
  final String id;

  /// Salon auquel le RDV est lié (critère d'acceptation #22).
  final String salonId;

  /// Coiffeur assigné, ou `null` (réservation au niveau salon, MVP #22).
  final String? hairdresserId;

  /// Jour du RDV (composantes de date seules ; repère UTC+0).
  final DateTime date;

  /// Heures de début/fin, format `HH:MM` (repère UTC+0).
  final String startTime;
  final String endTime;

  /// Statut du RDV — `pending` (« En attente ») à la création (#22).
  final AppointmentStatus status;

  /// Commentaire libre du client, ou `null`.
  final String? clientNote;

  /// Prestations réservées (≥ 1) — un RDV est lié à ≥ 1 prestation (#22).
  final List<BookedService> services;

  /// Vrai si le RDV est encore modifiable **par le client** (US-3.2, #23).
  ///
  /// Miroir **d'affichage** de la règle serveur `is_client_modifiable` : seuls les
  /// états actifs (`pending`/`confirmed`) sont modifiables ; un RDV terminé
  /// (`completed`) ou terminal (`cancelled`/`noShow`) est verrouillé côté client
  /// (§8.1). Aide UX uniquement — le serveur reste juge (un refus revient en `409`).
  bool get isClientModifiable =>
      status == AppointmentStatus.pending ||
      status == AppointmentStatus.confirmed;
}

File: web-dashboard/src/domain/appointment/appointment.ts
// Types & règles de domaine « rendez-vous » côté planning gérant (US-3.5, #26) —
// couche domaine (hexagonal, ADR-0008), TypeScript pur, testable sans React.
//
// Parité avec le backend (`coiflink_api/domain/appointment.py`) : l'union de
// statut reprend `AppointmentStatus` (§9.4) et les **prédicats d'action** sont le
// miroir de `ALLOWED_STATUS_TRANSITIONS` (machine à états gérant, #25). Ces
// prédicats ne font que **cacher** les boutons non pertinents : le backend reste
// l'**arbitre** (une transition interdite renvoie un `409`, traduit en message
// neutre côté UI). L'UI n'invente donc **aucune** transition.
//
// Aucun secret ni PII n'y figure. `priceAtBooking` est porté en **chaîne
// décimale** (parité `NUMERIC(12,2)`) pour ne pas perdre de précision.

export const APPOINTMENT_STATUSES = [
  "PENDING",
  "CONFIRMED",
  "CANCELLED",
  "COMPLETED",
  "NO_SHOW",
] as const;

export type AppointmentStatus = (typeof APPOINTMENT_STATUSES)[number];

export function isAppointmentStatus(value: string): value is AppointmentStatus {
  return (APPOINTMENT_STATUSES as readonly string[]).includes(value);
}

export interface BookedService {
  serviceId: string;
  // Montant décimal en chaîne (parité `NUMERIC(12,2)`), p. ex. "5000.00".
  priceAtBooking: string;
}

export interface Appointment {
  id: string;
  salonId: string;
  clientId: string;
  hairdresserId: string | null;
  // Jour du RDV, ISO "YYYY-MM-DD".
  date: string;
  // Heures locales du salon (Africa/Abidjan), "HH:MM:SS".
  startTime: string;
  endTime: string;
  status: AppointmentStatus;
  clientNote: string | null;
  services: BookedService[];
}

// Libellés **francisés** affichés (en attente | confirmé | annulé | terminé | absent).
export const STATUS_LABELS_FR: Record<AppointmentStatus, string> = {
  PENDING: "En attente",
  CONFIRMED: "Confirmé",
  CANCELLED: "Annulé",
  COMPLETED: "Terminé",
  NO_SHOW: "Absent",
};

// Classes Tailwind **littérales** par statut (jetons cohérents avec l'existant :
// `service-list.tsx` utilise `bg-palm/10 text-palm`, `bg-danger/10 text-danger`).
// Les chaînes sont écrites en entier (pas d'interpolation) pour rester détectables
// par le JIT Tailwind v4.
export interface StatusStyle {
  // Pastille pleine (agenda semaine, points de la grille mois).
  dot: string;
  // Badge/section coloré (vue jour, chips de légende).
  badge: string;
}

export const STATUS_STYLES: Record<AppointmentStatus, StatusStyle> = {
  PENDING: { dot: "bg-gold", badge: "border-gold/30 bg-gold/10 text-gold" },
  CONFIRMED: { dot: "bg-palm", badge: "border-palm/30 bg-palm/10 text-palm" },
  CANCELLED: { dot: "bg-danger", badge: "border-danger/30 bg-danger/10 text-danger" },
  COMPLETED: { dot: "bg-accent", badge: "border-accent/30 bg-accent/10 text-accent" },
  NO_SHOW: {
    dot: "bg-terracotta",
    badge: "border-terracotta/30 bg-terracotta/10 text-terracotta",
  },
};

// Statuts **terminaux** : aucune action de pilotage (miroir de `TERMINAL_STATUSES`).
export function isTerminal(status: AppointmentStatus): boolean {
  return status === "CANCELLED" || status === "COMPLETED" || status === "NO_SHOW";
}

// Prédicats d'action **miroir** de `ALLOWED_STATUS_TRANSITIONS` (#25) :
//   PENDING   → CONFIRMED (confirmer) | CANCELLED (refuser) | NO_SHOW (absent) ;
//   CONFIRMED → COMPLETED (terminer)  | NO_SHOW (absent)    | CANCELLED (annuler).
export function canConfirm(status: AppointmentStatus): boolean {
  return status === "PENDING";
}

export function canRefuse(status: AppointmentStatus): boolean {
  return status === "PENDING";
}

export function canComplete(status: AppointmentStatus): boolean {
  return status === "CONFIRMED";
}

export function canMarkNoShow(status: AppointmentStatus): boolean {
  return status === "PENDING" || status === "CONFIRMED";
}

export function canCancel(status: AppointmentStatus): boolean {
  return status === "CONFIRMED";
}

export type ActionTone = "primary" | "neutral" | "danger";

export interface StatusAction {
  // Statut cible d'une transition **autorisée** par l'état courant (#25).
  target: AppointmentStatus;
  label: string;
  tone: ActionTone;
}

// Actions proposées par état — strict sous-ensemble des transitions autorisées par
// le backend (deny-by-default) : un état terminal n'en propose **aucune**.
const TRANSITION_ACTIONS: Record<AppointmentStatus, readonly StatusAction[]> = {
  PENDING: [
    { target: "CONFIRMED", label: "Confirmer", tone: "primary" },
    { target: "CANCELLED", label: "Refuser", tone: "danger" },
    { target: "NO_SHOW", label: "Absent", tone: "neutral" },
  ],
  CONFIRMED: [
    { target: "COMPLETED", label: "Terminer", tone: "primary" },
    { target: "NO_SHOW", label: "Absent", tone: "neutral" },
    { target: "CANCELLED", label: "Annuler", tone: "danger" },
  ],
  CANCELLED: [],
  COMPLETED: [],
  NO_SHOW: [],
};

export function availableActions(status: AppointmentStatus): readonly StatusAction[] {
  return TRANSITION_ACTIONS[status] ?? [];
}
